stock_picking_batch_creation/wizards/make_picking_batch.py

```diff
@@ -0,0 +1,312 @@
+        first_picking.nbr_picking_lines = self._picking_nbr_lines(first_picking)
+        volume_per_bin = device.max_volume / device.nbr_bins
+        first_picking.nbr_bins_batch_picking = math.ceil(first_picking.total_volume_batch_picking / volume_per_bin)
+        is_overweighted = tools.float_compare(first_picking.total_weight_batch_picking, device.max_weight, 1) == 1
+        is_oversized = tools.float_compare(first_picking.total_volume_batch_picking, device.max_volume, 1) == 1
+        is_big_order = (
+            tools.float_compare(
+                first_picking.nbr_picking_lines, self.maximum_number_of_preparation_lines, 1
+            )
+            == 1
+        )
+        is_device_full = first_picking.nbr_bins_batch_picking > device.nbr_bins
+        if is_overweighted or is_oversized or is_big_order or is_device_full:
+            params = (first_picking.total_weight_batch_picking, first_picking.total_volume_batch_picking, first_picking.nbr_picking_lines, first_picking.nbr_bins_batch_picking)
+            msg = "The first picking does not match the device you wanted to use"
+            msg2 = "First picking needs: weight %s, volume %s, number of picking lines %s, number of bins %s" % params
+            raise UserError(_(msg+"\n"+ msg2))
+        
+        return first_picking, volume_per_bin
+
+    def _assign_operator_on_selected_pickings(self, batch, user):
+        batch.picking_ids.write({"user_id": user.id})
+
+    def _precision_weight(self):
+        return self.env["decimal.precision"].precision_get("Product Unit of Measure")
+
+    def _precision_volume(self):
+        return max(
+            6,
+            self.env["decimal.precision"].precision_get("Product Unit of Measure") * 2,
+        )
+
+    def _apply_limits(self, pickings, device, max_pickings=0):
+        """
+        Once we have the candidates pickings for clustering, we need to decide when to stop
+        the cluster (i.e.,  when to stop adding pickings in it)
+        1) If we only have one bin available, only one picking is selected
+        because we do not mix pickings in bins
+        2) If the first candidate picking for clustering is breaking one condition
+        (too voluminous or too heavy or too many lines in the picking or takes all the available bins)
+        3) We add pickings in the cluster as long as the conditions are respected
+        """
+        selected_pickings = self.env["stock.picking"].browse()
+        unselected_pickings = self.env["stock.picking"].browse()
+
+        first_picking = fields.first(pickings)
+        # 1) Check the number of bins available to create the cluster
+        selected_pickings = self._check_number_of_available_bins(
+            first_picking, device
+        )
+        if selected_pickings:
+            return selected_pickings, unselected_pickings
+        # 2) Check weither the first picking is breaking one condition or not
```

Comment:
@lmarion-source Why do you stop on first picking it this one is not suitable for the batch may be the next one could be use... We should iterate over the pickings to find the first one and only if we don't find a picking launch an expeption...